### ISSUE TITLE
Frontend preview throws exception for pages with `tl_page.requireItem`

### ISSUE THREAD

**Issue by SeverinGloeckle** [contao/contao#3525]
**Affected version(s)**

Contao 4.12, 4.11 (likely Contao 4.10 as well)

**Description**

Starting the frontend preview for a page with `tl_page.requireItem` enabled ("Require an item"-option) throws:

````php
Contao\CoreBundle\Exception\RouteParametersException

"Parameter "parameters" for route "cmf_routing_object" must match "/.+" ("" given) 
to generate a corresponding URL."
````

This exception is not caught.

**Details**
The exception is thrown since the router cannot generate the URL for a page which requires an item without an item parameter (somewhat similar to #3450).

**Possible fix**
Generating a proper URL is not possible in this case since there is no way to guess the correct item for the page. 

So there are two options left:
1. Generate the base URL of the page without the item parameter. This will likely show a 404 page in the frontend preview as a result of calling the page without an item parameter (similar to the old behaviour where modules like news readers would trigger the 404 on such pages).
2. Disable the frontend preview for pages which require an item and redirect to the default root page (that's how it is done for non-previewable content already)

Neither option is perfect, but each is preferable to the exception.

**Comment by fritzmg** [comment]
The preview will work if you launch the preview from within whichever item can be displayed on that page (e.g. a news article which is the child of a news archive that has said page defined as the redirect page).

**Comment by SeverinGloeckle** [comment]
That's correct. 

In this case the preview is generated by a custom URL converter which depends on the item (e.g. the `PreviewUrlConvertListener` in the news-bundle). That works within the context of the item's module only.

Starting the preview from the page structure or from within an article for pages with `tl_page.requireItem` should not throw an (unhandled) exception.

**Comment by aschempp** [comment]
You should get [this](https://github.com/contao/contao/pull/2019) custom exception if you try to generate the URL for the frontend preview. Is that not the case?

**Comment by SeverinGloeckle** [comment]
Yes, a `Contao\CoreBundle\Exception\RouteParametersException` is thrown as decribed above.

However, this exception / the resulting error screen should not be visible to the user if the frontend preview is started for a page that requires an item.

The linked pull request #3540 contains a fix by swallowing the exception in `PreviewUrlConvertListener`. 

Maybe you can share your opinion on whether the `RouteParametersException` should be swallowed in general or only for pages which require an item, see https://github.com/contao/contao/pull/3540/files#r721105391

**Comment by aschempp** [comment]
why should the error screen not be visible? A front end preview always requires a back end login.

**Comment by SeverinGloeckle** [comment]
I guess we have been talking past each other here.

Until now, in the specific case when a frontend preview is started for a page with `tl_page.requireItem`, an exception is thrown and an error screen is displayed.

Instead, the frontend preview should simply be generated with an empty URL (as in #3540).

Since you commented on the pull request regarding swallowing the exception, it's all cleared up.

This issue will be fixed once #3540 is merged.
